src/Parser/GuardianDruid/Modules/Features/RageWasted.js

```diff
@@ -0,0 +1,132 @@
+import React from 'react';
+import { formatPercentage } from 'common/format';
+import SPELLS from 'common/SPELLS';
+import RESOURCE_TYPES from 'common/RESOURCE_TYPES';
+import SpellLink from 'common/SpellLink';
+import SpellIcon from 'common/SpellIcon';
+import StatisticBox from 'Main/StatisticBox';
+import Module from 'Parser/Core/Module';
+
+// NOTE: "Raw" rage is what shows up in combat log events (divided by 10 and rounded to get in-game rage).
+// We deal with raw rage here to prevent accuracy loss.
+
+// This is actually ~59.95, for some bizarre reason, meaning that occasionally you will gain 5 rage
+// instead of 6 from a melee.  We have no idea why this is. As long as we re-sync with the classResources
+// field with each event, this shouldn't be a problem.
+const RAW_RAGE_GAINED_FROM_MELEE = 60;
+
+// The cost of a Gory Fur-reduced Ironfur, or the minimum amount of rage that can be spent
+const MINIMUM_ACCEPTABLE_RAGE_WASTE = 34;
+
+const RAGE_GENERATORS = {
+  [SPELLS.MELEE.id]: 'Melee',
+  [SPELLS.MANGLE_BEAR.id]: 'Mangle',
+  [SPELLS.THRASH_BEAR.id]: 'Thrash',
+  [SPELLS.MOONFIRE.id]: 'Moonfire (Galactic Guardian)',
+  [SPELLS.BLOOD_FRENZY_TICK.id]: 'Blood Frenzy',
+  [SPELLS.BRISTLING_FUR.id]: 'Bristling Fur',
+  [SPELLS.OAKHEARTS_PUNY_QUODS_BUFF.id]: 'Oakheart\'s Puny Quods',
+  [SPELLS.PURE_RAGE_POTION.id]: 'Pure Rage Potion',
+};
+
+class RageWasted extends Module {
+  rageWastedBySpell = {};
+  totalRageGained = 0;
+  currentRawRage = 0;
+
+  // Currently always 1000, but in case a future tier set/talent/artifact trait increases this it should "just work"
+  currentMaxRage = 0;
+
+  synchronizeRage(event) {
+    if (!event.classResources) {
+      console.log('no classResources', event);
+      return;
+    }
+    const rageResource = event.classResources.find(resource => resource.type === RESOURCE_TYPES.RAGE);
+    if (rageResource) {
+      this.currentRawRage = rageResource.amount;
+      this.currentMaxRage = rageResource.max;
+    }
+  }
+
+  registerRageWaste(abilityID, waste) {
+    if (!this.rageWastedBySpell[abilityID]) {
+      this.rageWastedBySpell[abilityID] = waste;
+    } else {
+      this.rageWastedBySpell[abilityID] += waste;
+    }
+  }
+
+  on_byPlayer_energize(event) {
+    this.synchronizeRage(event);
+    if (event.resourceChangeType !== RESOURCE_TYPES.RAGE) {
+      return;
+    }
+
+    if (event.waste > 0) {
+      this.registerRageWaste(event.ability.guid, event.waste);
+      this.totalRageGained += event.resourceChange + event.waste;
+    }
+  }
+
+  on_byPlayer_cast(event) {
+    if (event.ability.guid === SPELLS.MELEE.id) {
+      if (this.currentRawRage + RAW_RAGE_GAINED_FROM_MELEE > this.currentMaxRage) {
+        const realRageWasted = Math.round((this.currentRawRage + RAW_RAGE_GAINED_FROM_MELEE - this.currentMaxRage) / 10);
+        this.registerRageWaste(event.ability.guid, realRageWasted);
```

Comment:
Possibly. 

You can see this in the events themselves:
```
{
 "type":"energize",
 // These values are in "real rage"
 "resourceChange":4, 
 "resourceChangeType":1,
 "otherResourceChange":0,
 "waste":0,
 "classResources": [
  {
   // These values are in "raw rage"; divide by 10 to get the "real rage" amount
   "amount":410,
   "max":1000,
   "type":1
  }
 ],
...
}
```

 In an ideal scenario this is never an issue (all rage generators except Melee attacks generate a multiple-of-10 amount of raw rage), it's just that for some bizarre reason you can end up with a non-multiple-of-10 amount after a Melee attack.  

I investigated this extensively when I discovered that you occasionally only gain 5 rage instead of 6; on average, one in every ~20 melee attacks generates 59 raw rage instead of 60, and whenever this happens the value is floored to give you a real rage gain of only 5 (actually this reminds me that the Math.round should be changed to Math.floor).

Anyway long story short, the only reason I need to deal with raw rage at all is because of this anomaly where you can end up with 59 instead of 60 raw rage gain from melee attacks.